platform/source/mbed_retarget.cpp

```diff
@@ -1191,10 +1191,12 @@ extern "C" int fcntl(int fildes, int cmd, ...)
             va_start(ap, cmd);
             int flags = va_arg(ap, int);
             va_end(ap);
-            int ret = fhc->set_blocking(flags & O_NONBLOCK);
-            if (ret < 0) {
-                errno = -ret;
-                return -1;
+            if (flags & O_NONBLOCK) {
+                int ret = fhc->set_blocking(false);
+                if (ret < 0) {
+                    errno = -ret;
+                    return -1;
+                }
```

Comment:
I think the fix is not sufficient as it doesn't cover the case where the fd must be set to blocking. 
What do you think of: 

```suggestion
            int ret = fhc->set_blocking(!(flags & O_NONBLOCK));
            if (ret < 0) {
                errno = -ret;
                return -1;
```